app/controllers/posts_controller.rb

```diff
@@ -160,6 +160,67 @@ def show
 
   def edit; end
 
+  # Attempts to update a given post
+  # @param post [Post] post the user is attempting to update
+  # @param user [User] user attempting to update the post
+  # @param edit_post_params [ActionController::Parameters] edit parameters
+  # @param body_rendered [String] new post body
+  # @return [Boolean] status of the operation
+  def do_update(post, user, edit_post_params, body_rendered)
+    post.update(edit_post_params.merge(body: body_rendered,
+                                       last_edited_at: DateTime.now,
+                                       last_edited_by: user,
+                                       last_activity: DateTime.now,
+                                       last_activity_by: user))
+  end
+
+  # Attempts to update a given post network-wide. The update is manual to avoid
+  # skipping validations and fail early if at least one validation fails.
+  # @param post [Post] post from which the network push is initiated
+  # @param posts [ActiveRecord::Result] network posts to be updated
+  # @param user [User] user attempting to push updates to network
+  # @param edit_post_params [ActionController::Parameters] edit parameters
+  # @param body_rendered [String] new post body
+  # @return [Boolean] status of the operation
+  def do_update_network(post, posts, user, edit_post_params, body_rendered)
+    update_status = true
+
+    posts.each do |network_post|
+      network_post.update(edit_post_params.merge(body: body_rendered,
+                                                 last_edited_at: DateTime.now,
+                                                 last_edited_by_id: user.id,
+                                                 last_activity: DateTime.now,
+                                                 last_activity_by_id: user.id))
+
+      if network_post.errors.any?
+        post.errors.merge!(network_post.errors)
+        update_status = false
+      end
+
+      next if update_status == true
+    end
+
+    update_status
+  end
+
+  # Checks if a given user can push a given post type to network
+  # @param user [User] user attempting to push to network
+  # @param post_type [PostType] type of the post to be pushed
+  # @return [Boolean]
+  def can_push_to_network(user, post_type)
+    post_type.system? && (user.is_global_moderator || user.is_global_admin)
+  end
+
+  # Checks if a given user can directly update a given post
+  # @param post [Post] updated post (owners can unilaterally update)
+  # @param user [User] user attempting to update the post
+  # @param post_type [PostType] type of the post (some are freely editable)
+  # @return [Boolean]
+  def can_update(post, user, post_type)
+    user.privilege?('edit_posts') || user.is_moderator || user == post.user || \
+      (post_type.is_freely_editable && user.privilege?('unrestricted'))
+  end
```

Comment:
suggestion: these should probably be methods on the `Post` or `User` model rather than in a controller. The `do_update` methods could as well, to be fair.

Reply:
Yup, they should definitely not be here, an artifact of abstracting the logic from the `update` method (otherwise, with the changes it starts to fail the cyclomatic complexity rule of RuboCop [which is fair, the method is a heavy one]). Done and done, with a couple of unit tests as well.

Reply:
That said, I kept the `do_update` methods on the controller _for now_ as I am a bit worried about shuffling it around right now. I'd like to tackle the move separately later as I also have a couple of other improvements in mind (trying not to scope creep this one).